### ISSUE TITLE
Error building custom recipes and no receipe-file found

### ISSUE THREAD

**Issue by schattenlink** [crops/toaster-container#21]
Hello,
If I try to build a custom image it ends up after 3 seconds with "Error: BRLayer matching query does not exist."

And if I try to download any recipe I get the Django Error page

Django Version: 	1.8.15
Exception Type: 	IOError
Exception Value: 	

[Errno 2] No such file or directory: u'recipes-core/images/core-image-minimal.bb'

Exception Location: 	/home/usersetup/poky/bitbake/lib/toaster/orm/models.py in generate_recipe_file_contents, line 1613

The Debug Infos said:

ERROR runbuilds: Error launching build BRLayer matching query does not exist.
Traceback (most recent call last):
  File "/home/usersetup/poky/bitbake/lib/toaster/bldcontrol/management/commands/runbuilds.py", line 67, in schedule
    "%d:%d" % (br.pk, bec.be.pk))
  File "/home/usersetup/poky/bitbake/lib/toaster/bldcontrol/localhostbecontroller.py", line 255, in triggerBuild
    layers = self.setLayers(bitbake, layers, targets)
  File "/home/usersetup/poky/bitbake/lib/toaster/bldcontrol/localhostbecontroller.py", line 211, in setLayers
    layer_version=customrecipe.base_recipe.layer_version)
  File "/usr/local/lib/python2.7/dist-packages/django/db/models/query.py", line 334, in get
    self.model._meta.object_name
DoesNotExist: BRLayer matching query does not exist.

I just pulled the Container, and run it with the mounted empty workdir.

No further configuration or so.

Please help me.

**Comment by bavery22** [comment]
We'll certainly try to help :) Need some more information if you could
Was this container crops/toaster:latest?
What layers did you add, if any?
What builds did you do before building the custom image and did they succeed?
What machine were you targeting?
Your project was based on what type? Local? Morty? ,,,

Basically, I'd like to recreate the bug...

**Comment by schattenlink** [comment]
The Container is just crops/toaster. If it isn't nearly described I think it automately takes "latest".
I build core-image-minimal for qemux86 successfully and add no layer to it or the custom one.
Toaster offers only Jethro and Krogoth. There is no Morty.
I used both Jethro and Krogoth with the same outcome like described above.

Many Thanks for your Help.

**Comment by bavery22** [comment]
Hi,

I tried and failed to reproduce this today :(.

I was on a linux host (Ubuntu 14.04) with docker 1.12.3.

What I did:
$ docker pull crops/toaster
$ mkdir workspace
$ docker run -it --rm -p 0.0.0.0:18000:8000 -v `pwd`/workspace:/workdir
crops/toaster --workdir=/workdir

Then I made a project -> Krogoth -> qemux86
Recipe to build -> core-image-minimal (success)
New Custom Image - c1 - based on core-image-minimal
Build C1 Button  (success).

Then I went to Software recipes and added tcpdump from meta-networking.
Built tcpdump (success)
Made another custom image - c2, based on core-image-minimal
added the tcpdump package to the c2 image.
Built the c2 image (success).
------

So, a couple of questions:
1) what host os are you on?
2) could you attach the results of docker inspect <container id for toaster>
3) what command line did you run with?
4) did you build core-image-minmal and then the custom image w/out
restarting the container?


-brian
an intel employee

**Comment by schattenlink** [comment]
I am on a Debian Jessie VM with Docker 1.12.4

Docker inspect:

`$ docker inspect tender_lovelace
[
    {
        "Id": "5d77ffe128aaa958efdcdfaf64af4f7b524647e9e630364d660de8dc330ff955",
        "Created": "2017-01-05T11:13:49.394761306Z",
        "Path": "toaster-entry.py",
        "Args": [],
        "State": {
            "Status": "running",
            "Running": true,
            "Paused": false,
            "Restarting": false,
            "OOMKilled": false,
            "Dead": false,
            "Pid": 12843,
            "ExitCode": 0,
            "Error": "",
            "StartedAt": "2017-01-05T11:13:50.459509064Z",
            "FinishedAt": "0001-01-01T00:00:00Z"
        },
        "Image": "sha256:c63d62077603ec5eea71cb9adb8876785e63bf4c7c8d97e63ce7d50c355d8975",
        "ResolvConfPath": "/var/lib/docker/containers/5d77ffe128aaa958efdcdfaf64af4f7b524647e9e630364d660de8dc330ff955/resolv.conf",
        "HostnamePath": "/var/lib/docker/containers/5d77ffe128aaa958efdcdfaf64af4f7b524647e9e630364d660de8dc330ff955/hostname",
        "HostsPath": "/var/lib/docker/containers/5d77ffe128aaa958efdcdfaf64af4f7b524647e9e630364d660de8dc330ff955/hosts",
        "LogPath": "/var/lib/docker/containers/5d77ffe128aaa958efdcdfaf64af4f7b524647e9e630364d660de8dc330ff955/5d77ffe128aaa958efdcdfaf64af4f7b524647e9e630364d660de8dc330ff955-json.log",
        "Name": "/tender_lovelace",
        "RestartCount": 0,
        "Driver": "aufs",
        "MountLabel": "",
        "ProcessLabel": "",
        "AppArmorProfile": "",
        "ExecIDs": [
            "01802d31c2cbc7278a2ab7f22d1ef379f14bd08a660ab55773cb1361ed227338",
            "b1185ac10e0eab1778ecca9be022664cc8217fa2d40e2b37060a3f2e538ab803"
        ],
        "HostConfig": {
            "Binds": [
                "/home/ritter/toasterstuff:/workdir"
            ],
            "ContainerIDFile": "",
            "LogConfig": {
                "Type": "json-file",
                "Config": {}
            },
            "NetworkMode": "default",
            "PortBindings": {
                "8000/tcp": [
                    {
                        "HostIp": "",
                        "HostPort": "8000"
                    }
                ]
            },
            "RestartPolicy": {
                "Name": "no",
                "MaximumRetryCount": 0
            },
            "AutoRemove": false,
            "VolumeDriver": "",
            "VolumesFrom": null,
            "CapAdd": null,
            "CapDrop": null,
            "Dns": [],
            "DnsOptions": [],
            "DnsSearch": [],
            "ExtraHosts": null,
            "GroupAdd": null,
            "IpcMode": "",
            "Cgroup": "",
            "Links": null,
            "OomScoreAdj": 0,
            "PidMode": "",
            "Privileged": false,
            "PublishAllPorts": false,
            "ReadonlyRootfs": false,
            "SecurityOpt": null,
            "UTSMode": "",
            "UsernsMode": "",
            "ShmSize": 67108864,
            "Runtime": "runc",
            "ConsoleSize": [
                0,
                0
            ],
            "Isolation": "",
            "CpuShares": 0,
            "Memory": 0,
            "CgroupParent": "",
            "BlkioWeight": 0,
            "BlkioWeightDevice": null,
            "BlkioDeviceReadBps": null,
            "BlkioDeviceWriteBps": null,
            "BlkioDeviceReadIOps": null,
            "BlkioDeviceWriteIOps": null,
            "CpuPeriod": 0,
            "CpuQuota": 0,
            "CpusetCpus": "",
            "CpusetMems": "",
            "Devices": [],
            "DiskQuota": 0,
            "KernelMemory": 0,
            "MemoryReservation": 0,
            "MemorySwap": 0,
            "MemorySwappiness": -1,
            "OomKillDisable": false,
            "PidsLimit": 0,
            "Ulimits": null,
            "CpuCount": 0,
            "CpuPercent": 0,
            "IOMaximumIOps": 0,
            "IOMaximumBandwidth": 0
        },
        "GraphDriver": {
            "Name": "aufs",
            "Data": null
        },
        "Mounts": [
            {
                "Source": "/home/ritter/toasterstuff",
                "Destination": "/workdir",
                "Mode": "",
                "RW": true,
                "Propagation": "rprivate"
            }
        ],
        "Config": {
            "Hostname": "5d77ffe128aa",
            "Domainname": "",
            "User": "usersetup",
            "AttachStdin": true,
            "AttachStdout": true,
            "AttachStderr": true,
            "ExposedPorts": {
                "8000/tcp": {}
            },
            "Tty": true,
            "OpenStdin": true,
            "StdinOnce": true,
            "Env": [
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
                "LANG=en_US.UTF-8"
            ],
            "Cmd": null,
            "Image": "crops/toaster",
            "Volumes": null,
            "WorkingDir": "/home/yoctouser",
            "Entrypoint": [
                "toaster-entry.py"
            ],
            "OnBuild": null,
            "Labels": {}
        },
        "NetworkSettings": {
            "Bridge": "",
            "SandboxID": "bd7a8f3919d42944b6aacf170d01afb10e1e761bc1848ec93e1554d52e17febb",
            "HairpinMode": false,
            "LinkLocalIPv6Address": "",
            "LinkLocalIPv6PrefixLen": 0,
            "Ports": {
                "8000/tcp": [
                    {
                        "HostIp": "0.0.0.0",
                        "HostPort": "8000"
                    }
                ]
            },
            "SandboxKey": "/var/run/docker/netns/bd7a8f3919d4",
            "SecondaryIPAddresses": null,
            "SecondaryIPv6Addresses": null,
            "EndpointID": "afb2600d51ec28fcc6670bb54047a6bc6b53096e968aa744d1405eeab4f27924",
            "Gateway": "172.17.0.1",
            "GlobalIPv6Address": "",
            "GlobalIPv6PrefixLen": 0,
            "IPAddress": "172.17.0.2",
            "IPPrefixLen": 16,
            "IPv6Gateway": "",
            "MacAddress": "02:42:ac:11:00:02",
            "Networks": {
                "bridge": {
                    "IPAMConfig": null,
                    "Links": null,
                    "Aliases": null,
                    "NetworkID": "2166950c80422138ec961764d91cdc9878ff9ce543bdd53bdcb60d21888b7b2f",
                    "EndpointID": "afb2600d51ec28fcc6670bb54047a6bc6b53096e968aa744d1405eeab4f27924",
                    "Gateway": "172.17.0.1",
                    "IPAddress": "172.17.0.2",
                    "IPPrefixLen": 16,
                    "IPv6Gateway": "",
                    "GlobalIPv6Address": "",
                    "GlobalIPv6PrefixLen": 0,
                    "MacAddress": "02:42:ac:11:00:02"
                }
            }
        }
    }
]



I run the container with:
`docker run -it --rm -p 8000:8000 -v /home/ritter/toasterstuff:/workdir crops/toaster`

Yes i build tried to build it without restarting.

**Comment by schattenlink** [comment]
I just reinitialize the container.... same result...

**Comment by rewitt1** [comment]
Hi, I just built a custom image using the container in a debian jessie vm running on a fedora-24 host. I couldn't reproduce your error either.

What happens if you start with an empty "toasterstuff" directory? Does it still happen?

You could also give us the output of "docker info", just in case there is something useful in there.

**Comment by schattenlink** [comment]
Hello,
I purged "toasterstuff" and rerun the container.

on trying to download the recipe for core-image-minimal I get:
Page not found (404)
Request Method: 	GET
Request URL: 	http://xxx.xx.xx.xxx:8000/toastergui/project/2/recipe/None 

Here comes the Docker-info but there is nothing interesting to be found:

$ docker info
Containers: 6
 Running: 1
 Paused: 0
 Stopped: 5
Images: 12
Server Version: 1.12.4
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 32
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: host bridge overlay null
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options:
Kernel Version: 3.16.0-4-amd64
Operating System: Debian GNU/Linux 8 (jessie)
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 44.62 GiB
Name: VM-SDFG
ID: 33JB:QFNI:5WJJ:5OGN:DIVQ:7GBT:GXBI:C3OO:4ZVR:7JYH:2HL5:V4G4
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No memory limit support
WARNING: No swap limit support
WARNING: No kernel memory limit support
WARNING: No oom kill disable support
WARNING: No cpu cfs quota support
WARNING: No cpu cfs period support
Insecure Registries:
 127.0.0.0/8

**Comment by schattenlink** [comment]
Where should the Recipe-File been created? in the toasterstuff there is no toastergui/project-Folder by initialisation or by creating a new project.

toasterstuff$ tree
.
└── build
    ├── conf
    │   ├── bblayers.conf
    │   ├── local.conf
    │   └── templateconf.cfg
    ├── pyshtables.py
    ├── toaster.sqlite
    └── toaster_web.log

**Comment by rewitt1** [comment]
I did just reproduce the error when clicking "Download Recipe", evidently the recipe isn't created until some portion of the build succeeds. So if you click "Download Recipe" before you try to do a build, or if the build fails early enough, "Download Recipe" will fail. Honestly, that's a bug and it shouldn't let you try if it's just going to fail. IF it did make it far enough the recipe would be in a layer under build/toaster-custom-images

So now we're still left with the question of why your build is failing. Is there any way you could do a screen recording of all the steps from creating the "toasterstuff" workdir, doing "docker pull crops/toaster:latest" as well as the docker run command? Followed by opening your browser and doing everything needed that causes it to fail for you?

I know this is a bit tedious, but since it is a UI, it will allow me to do "exactly" what you did. In debian you can hit Control+Shift+Alt+R to start a screencast recording, if you are using gnome. Here are the instructions that also detail where the saved recording will be: https://wiki.gnome.org/Projects/GnomeShell/CheatSheet#Screencast_recording

Once you have the recording, please upload it here or wherever is convenient and point us at it. Thanks.

**Comment by schattenlink** [comment]
Hi, I cleaned toasterstuff, delete die docker containers and images, and reboot the debian VM.

[2-build.zip](https://github.com/crops/toaster-container/files/698924/2-build.zip)
[1-start.zip](https://github.com/crops/toaster-container/files/698925/1-start.zip)

after building core-image-minimal I also cant download its recipe-file
Page not found (404)
Request Method: 	GET
Request URL: 	http://xxx.xx.xxx.xxx:8000/toastergui/project/2/recipe/None

Thanks for helping me

in between: there are hundreds of zombie-processes left after building the image.

**Comment by rewitt1** [comment]
I finally reproduced this! It is definitely a bug in toaster that we'll put into bugzilla.

However, we were successfully building custom images by taking a different route. So while that shouldn't be the case, it may let you try out what you wanted.

I've attached a screenshot. It is the screen that is shown right after creating a new project. On it you will see "New Custom Image".  If you use that button to create a new image after building core-image-minimal, instead of the route you were taking, it should work.

<img width="1392" alt="screen shot 2017-01-11 at 6 57 33 pm" src="https://cloud.githubusercontent.com/assets/3660987/21876008/bf0805e8-d834-11e6-88f5-65c73d3f05ef.png">

Regarding the zombie process, it's a known issue and why https://bugzilla.yoctoproject.org/show_bug.cgi?id=10530 exists.

One more thing, if you did want to use morty instead, it is available as crops/toaster-morty. There was a bug preventing us from moving crops/toaster:latest to morty, but that should happen soon.
